pkg/database/token.go

```diff
@@ -336,23 +336,57 @@ func (db *Database) updateStatsCodeInvalid(t time.Time, authApp *AuthorizedApp)
 				SET codes_invalid = authorized_app_stats.codes_invalid + 1
 		`
 	if err := db.db.Exec(authAppSQL, t, authApp.ID).Error; err != nil {
-		db.logger.Errorw("failed to update authorized app stats", "error", err)
+		db.logger.Errorw("failed to update authorized app stats code invalid", "error", err)
 	}
 }
 
 // updateStatsCodeClaimed updates the statistics, increasing the number of codes
 // claimed.
-func (db *Database) updateStatsCodeClaimed(t time.Time, authApp *AuthorizedApp) {
-	t = timeutils.UTCMidnight(t)
+func (db *Database) updateStatsCodeClaimed(t time.Time, authApp *AuthorizedApp, vc *VerificationCode) {
+	midnight := timeutils.UTCMidnight(t)
 
-	realmSQL := `
-			INSERT INTO realm_stats(date, realm_id, codes_claimed)
-				VALUES ($1, $2, 1)
-			ON CONFLICT (date, realm_id) DO UPDATE
-				SET codes_claimed = realm_stats.codes_claimed + 1
-		`
-	if err := db.db.Exec(realmSQL, t, authApp.RealmID).Error; err != nil {
-		db.logger.Errorw("failed to update realm stats", "error", err)
+	if err := db.db.Transaction(func(tx *gorm.DB) error {
+		var existing RealmStat
+		if err := tx.
+			Set("gorm:query_option", "FOR UPDATE").
+			Table("realm_stats").
+			Where("realm_id = ?", authApp.RealmID).
+			Where("date = ?", midnight).
+			Take(&existing).
+			Error; err != nil && !IsNotFound(err) {
+			return err
+		}
+
+		existing.CodesClaimed++
+		if len(existing.CodeClaimDistribution) == 0 {
+			existing.CodeClaimDistribution = make([]int32, len(claimDistributionBuckets))
+		}
+
+		sinceIssue := t.Sub(vc.IssuedAt)
```

Comment:
We could try avoid the missing data by instead trying to subtract from expiry (assuming it hasn't changed) but adding explicit IssuedAt feels more foolproof